cc/algorithms/partition-selection.h

```diff
@@ -0,0 +1,327 @@
+    double rand_num = UniformDouble();
+    //only keep partition if random number < expected probability of keep
+    return (rand_num <= ProbabilityOfKeep(num_users));
+  }
+
+ private:
+  double adjusted_epsilon_;
+  double crossover_1_;
+  double crossover_2_;
+
+  // ProbabilityOfKeep returns the probability with which a partition with n
+  // users should be kept, Thm. 1 of https://arxiv.org/pdf/2006.03684.pdf
+  double ProbabilityOfKeep(double n) {
+    double adjusted_delta_ = GetAdjustedDelta();
+    if(n == 0) {
+      return 0;
+    }
+    else if (n <= crossover_1_) {
+      return (((exp(n * adjusted_epsilon_) - 1) / (exp(adjusted_epsilon_) - 1))
+              * adjusted_delta_);
+    }
+    else if (n > crossover_1_ && n <= crossover_2_) {
+      double m = n - crossover_1_;
+      return ((1 - exp(-1 * m * adjusted_epsilon_))
+               * (1 + adjusted_delta_ / (exp(adjusted_epsilon_) - 1))
+               + exp(-1 * m * adjusted_epsilon_)
+               * ProbabilityOfKeep(crossover_1_));
+    }
+    else {
+      return 1;
+    }
+  }
+
+};
+
+// LaplacePartitionSelection calculates a threshold based on the CDF of the
+// Laplace distribution, delta, epsilon, and the max number of partitions a
+// single user can contribute to. If the number of users in a partition
+// + Laplace noise is greater than this threshold, the partition should be kept.
+class LaplacePartitionSelection : public PartitionSelectionStrategy {
+ public:
+  //Builder for LaplacePartitionSelection
+  class Builder : public PartitionSelectionStrategy::Builder {
+   public:
+     Builder& SetLaplaceMechanism(
+       std::unique_ptr<LaplaceMechanism::Builder> laplace_builder) {
+       laplace_builder_ = std::move(laplace_builder);
+       return *this;
+     }
+
+     base::StatusOr<std::unique_ptr<PartitionSelectionStrategy>>
+     Build() override {
+       RETURN_IF_ERROR(MaxPartitionsContributedIsSetAndValid());
+       RETURN_IF_ERROR(DeltaIsSetAndValid());
+       if (!GetEpsilon().has_value()) {
+         return base::InvalidArgumentError("Epsilon has to be set.");
+       }
+       if (!laplace_builder_.has_value()) {
+        laplace_builder_ = absl::make_unique<LaplaceMechanism::Builder>();
+       }
+       std::unique_ptr<NumericalMechanism> mechanism_;
+       ASSIGN_OR_RETURN(mechanism_,
+                        laplace_builder_.value()->
+                        SetEpsilon(GetEpsilon().value())
+                        .SetL0Sensitivity(GetMaxPartitionsContributed().value())
+                        .SetLInfSensitivity(1).Build());
+       std::unique_ptr<PartitionSelectionStrategy> laplace =
+         absl::make_unique<LaplacePartitionSelection>(
+           GetEpsilon().value(), GetDelta().value(),
+           GetMaxPartitionsContributed().value(), std::move(mechanism_));
+       return laplace;
+     }
+
+   private:
+     absl::optional<std::unique_ptr<LaplaceMechanism::Builder>> laplace_builder_;
```

Comment:
You basically never need an optional<unique_ptr<X>> because the unique_ptr can be nullptr, which already stands for "doesn't have a value".